custom_components/sonnenbatterie/sensor.py

```diff
@@ -61,517 +58,68 @@ def _internal_setup(_username, _password, _ipaddress):
+        self.coordinator = coordinator
+        self.entity_description = entity_description
 
-        calc_remaining_capacity_usable = max(
-            0, int(calc_remaining_capacity - calc_reserved_capacity)
+        self._attr_device_info = coordinator.device_info
+        self._attr_translation_key = (
+            tkey
+            if (tkey := entity_description.translation_key)
+            else entity_description.key
         )
+        if precision := entity_description.suggested_display_precision:
+            self._attr_suggested_display_precision = precision
 
-        sensor_name = "{}{}".format(
-            self.allSensorsPrefix, "state_remaining_capacity_usable"
-        )
-        unit_name = "Wh"
-        friendly_name = "Remaining Capacity Usable"
-        self._add_or_update_entity(
-            sensor_name,
-            friendly_name,
-            calc_remaining_capacity_usable,
-            unit_name,
-            SensorDeviceClass.ENERGY,
+        self.entity_id = (
+            f"sensor.sonnenbatterie_{self.coordinator.serial}_{self.entity_description.key}"
         )
 
-        """powermeter values"""
-        for meter in self.latestData["powermeter"]:
-            sensor_name_prefix = "{}meter_{}_{}_{}".format(
-                self.allSensorsPrefix,
-                meter["direction"],
-                meter["deviceid"],
-                meter["channel"],
-            )
-            sensor_name_prefix = sensor_name_prefix.lower()
-            generate_sensors_for = {
-                "a_l1",
-                "a_l2",
-                "a_l3",
-                "v_l1_l2",
-                "v_l1_n",
-                "v_l2_l3",
-                "v_l2_n",
-                "v_l3_l1",
-                "v_l3_n",
-                "w_l1",
-                "w_l2",
-                "w_l3",
-                "w_total",
-            }
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return f"{self.coordinator.serial}-{self.entity_description.key}"
 
-            for sensor_meter in generate_sensors_for:
-                sensor_name = "{}_{}".format(sensor_name_prefix, sensor_meter)
-                val = round(meter[sensor_meter], 2)
-                unit_name = (sensor_meter[0] + "").upper()
-                device_class = SensorDeviceClass.POWER
-                if unit_name == "V":
-                    device_class = SensorDeviceClass.VOLTAGE
-                elif unit_name == "A":
-                    device_class = SensorDeviceClass.CURRENT
-                friendly_name = "{0} {1}".format(meter["direction"], sensor_meter)
-                self._add_or_update_entity(
-                    sensor_name, friendly_name, val, unit_name, device_class
-                )
+        # legacy support / prevent breaking changes (will not directly work this way because keys changed)
+        # return f"sensor.sonnenbatterie_{self.coordinator.serial}_{self.entity_description.key}"
```

Comment:
Siehe `_atttr_has_entity_name` weiter oben ... 
Wenn geänderte Key-Namen unumgänglich sind, dann müssten wir da zumindest eine Liste mitliefern. Sonst denken die Leute nach dem Update ihre Integration ist kaputt.

Reply:
Wie gesagt sind sie nicht unumgänglich, aber es wäre halt besser wenn sie einheitlicher benannt werden.
Wir können auch weiterhin die alten nehmen, dann muss ich die halt wieder alle umbenennen (in der Entity Liste und in den translations).

Reply:
Naja, ich gehe halt davon aus, dass es Leute gibt, die bereits die alte Integration benutzen und ggf. darauf aufbauend auch Automatisierungen gebastelt haben. Da aus meiner Erfahrung **niemand** das Readme liest, sondern einfach nur auf Update geklickt wird und der nächste Weg dann sofort der zum Issue-Tracker ist, weil die Anzeige/Automatisierung/was-auch-sonst-immer nicht geht, denke ich es wäre netter gegenüber den Nutzern, nach Möglichkeit die alten Entity-Bezeichnungen zu behalten. Wenn es triftige _technische_ Gründe für eine Änderung gibt, dann ist das wieder was anderes.
Letztlich ist das aber nicht meine Entscheidung, das darf der Chef machen @weltmeyer ;)

Reply:
Was wir auch machen könnten wäre, für jeden Sensor ein weiteres Attribut `legacy_key' hinzuzufügen, das dann den alten key beinhalten würde, um den `unique_key` zu generieren um die existierenden entities zu behalten. Dadurch könnten wir intern weiterhin die neuen keys verwenden, die auch für die entity ids benutzt werden würden.
Und wer weiß, vielleicht gibt es ja irgendwann die Möglichkeit, unique keys zu migrieren.
Das würde eben ein paar zusätzliche Zeilen in der `SENSORS` Liste brauchen.

Reply:
> Und wer weiß, vielleicht gibt es ja irgendwann die Möglichkeit, unique keys zu migrieren. Das würde eben ein paar zusätzliche Zeilen in der `SENSORS` Liste brauchen.

Wenn ich mir https://community.home-assistant.io/t/replace-entities-and-keep-history-breaking-change-starting-ha-core-2023-4/639465 und https://community.home-assistant.io/t/how-to-replace-entity-in-energy-dashboard-by-new-one-without-losing-history/545933 so durchlese, dann fürchte ich, dass genau das jetzt nicht mehr geht. Für mich liest sich das so, als wären die alten `unique_ids` in Stein gemeisselt.

Reply:
Ich wäre sehr dafür, dass die entity_ids sich nicht verändern. Das würde doch meines Wissens nach nicht nur automatisierungen killen sondern auch alle entsprechenden historien, richtig?
![image](https://github.com/weltmeyer/ha_sonnenbatterie/assets/1668465/990bdbc7-86d3-4dd6-b1f6-d621f2e99ab1)
Ich glaube das wäre unschön, wenn die plötzlich nicht mehr passen.
die lösung mit dem legacy keys würde quasi bestehende User "absplitten" da diese mit den alten ID's arbeiten, bis mal aus irgendwelchen Gründen eine Neuinstallation ansteht, dann geht das dort auch kaputt...

Ich würde also gerne bei den alten Namen bleiben.
Eventuell an/abschaltbar "Neues Namensschema","Altes Namensschema" als Checkbox in der Konfig.. und dort dann auch beides gleichzeitig erlauben. Wir müssen dann halt ein Legacy<>New Mapping hinterlegen, Sensoren würden doppelt mit unterschiedlichen Namen vorhanden sein, wenn beide Häkchen aktiviert sind. Eventuell hinzukommende neue Sensoren ohne Legacy mapping werden egal was angehakt ist mit dem neuen Namen angelegt, da dieser ja bisher nicht in irgendeiner Automatisierung verwendet wurde.

Oh viel Prosa, sorry..

Reply:
Durch das Update werden die IDs ja nicht automatisch umbenannt. Das ist ja wie der Name, das bleibt bestehen. Deshalb hätte das dann auch keinen Einfluss auf die Statistiken. Es würde aber für neue Instllationen die neuen Entity IDs verwenden. Das sagst du ja eigentlich auch schon selbst weiter unten. Da meinst du ja, man würde die User dadurch absplitten, weil manche neue und manche alte Entity IDs haben.
Das würde ich so jetzt aber auch nicht sagen, denn man kann ja die Entity ID wie den Namen sowieso beliebig ändern.
Das habe ich bei mir z.B. mit den aktuellen Sensoren auch gemacht und wäre dadurch ja sozusagen auch "abgesplittet". Der Logik nach wäre man ja auch abgesplittet, wenn man die Entity einfach nur umbenennt.

Für den Fall einer Neuinstallation:
Falls man noch ein Backup hat und das einspielen möchte, werden ja auch die alten Entity IDs wieder verwendet.
Wenn man kein Backup mehr hat, muss man sowieso alles neu einrichten. Dann ist es auch egal, ob die Entity IDs sich geändert haben.
Für den Fall, dass man, warum auch immer, doch die alten Entity IDs weiterverwenden möchte (weil man vielleicht noch die Automatisierungen in einem Backup hat aber sonst nichts zum wiederherstellen) und diese nicht an das neue Schema anpassen will, kann man ja auch die Entity IDs wieder manuell in die alten umändern.

Das mit der Einstellung neues / altes schema finde ich jetzt nicht so gut, das macht es meiner Meinung nach noch uneinheitlicher und komplizierter, vorallem weil viele wahrscheinlich nicht verstehen werden, wozu die gut ist.

Reply:
Hmm also wenn durch das update als solches die Entity-IDs nicht verändert werden und die namen sich nur durch eine neu-installation verändern, dann würde ich das okay finden - bei einer Neuinstallation kann man ja auch gleich die eigenen Automatisierungen etc anpassen.. macht man ja auch nicht jeden Tag. Mir war das so nicht bewusst, dass die erhalten bleiben bis zur neu-Installation oder ähnliches.

Reply:
Aktuell spinnt meine Sonnenbatterie(Komme weder zuverlässig auf die API noch auf das lokale Dashboard... -.-), daher kann ich den PR nicht sinnvoll testen. @RustyDust hast du die Möglichkeit den PR als Update zu testen?

Reply:
> Aktuell spinnt meine Sonnenbatterie(Komme weder zuverlässig auf die API noch auf das lokale Dashboard... -.-),

Möglicherweise in letzter Zeit das LAN Kabel bewegt?
Meine Sonnenbatterie ging von Anfang an auch nur sehr unzuverlässig und langsam, bis ich irgendwann rausgefunden hab, dass die Techniker das Kabel ziemlich in die Kabelkanäle gequetscht haben und es dadurch kaputt gegangen ist. Mein Kabeltester hat wie wild durcheinander geblinkt anstatt in der Reihenfolge 1-8 :laughing:

Reply:
> Hmm also wenn durch das update als solches die Entity-IDs nicht verändert werden und die namen sich nur durch eine neu-installation verändern, dann würde ich das okay finden

Gut, dann werde ich das mit den `legacy_key`s machen.

Reply:
> @RustyDust hast du die Möglichkeit den PR als Update zu testen?

Wenn ich wieder Land sehe :/ Ich wollte das schon längst gemacht haben, aber die letzten Wochen waren die Hölle auf Arbeit ... ich schau wie ich dazu komme und melde mich.

Reply:
Okay, kein Problem. Ich glaube es wäre sowieso besser, wenn ich zuerst die `legacy_key`s mache, damit du dann das "finale" Update testen kannst.

Reply:
Habe die Option hinzugefügt und bis auf diese Sensoren wird nun alles übernommen, inklusive Namen und Entity IDs.
![ApplicationFrameHost_9YECvj50iC](https://github.com/weltmeyer/ha_sonnenbatterie/assets/59659167/9a06973e-7af1-488e-bdeb-5ae6db8e0e66)